src/core/presentation_definition.rs

```diff
@@ -263,6 +264,50 @@ impl PresentationDefinition {
 
         Ok(())
     }
+
+    /// Validate a JSON-encoded credential against the presentation definition.
+    ///
+    /// NOTE: this method accepts a generic serde_json::Value argument and checks whether
+    /// the JSON value conforms to the presentation definition's input descriptor constraint
+    /// fields.
+    ///
+    /// If the credential satisifies the presentation definition, this method will return true.
+    pub fn check_credential_validation(&self, credential: &serde_json::Value) -> bool {
+        let selector_lock = RwLock::new(jsonpath_lib::selector(credential));
+
+        self.input_descriptors().iter().any(|descriptor| {
+            descriptor
+                .constraints()
+                .fields()
+                .iter()
+                .filter_map(|field| match field.validator() {
+                    Some(Ok(validator)) => {
+                        let is_valid = field
+                            .path()
+                            .iter()
+                            .filter_map(|path| match selector_lock.try_write() {
+                                Ok(mut selector) => selector(path).ok(),
+                                _ => None,
+                            })
+                            .flatten()
+                            // NOTE: This is currently assuming that if any of the paths are a match
+                            // to the credential, then the validation is, at least partially, successful,
+                            // and the credential may satisfy the presentation definition.
+                            .any(|value| validator.validate(value).is_ok());
+
+                        // NOTE: if the field is not required and is not a match,
+                        // skip and continue checking the other fields.
+                        if !is_valid && field.is_optional() {
+                            return None;
+                        }
+
+                        Some(is_valid)
+                    }
+                    _ => None,
```

Comment:
The error case here is not handled. What should this library do if it receives a JSONSchema filter that cannot be parsed? Based on the [specification](https://openid.net/specs/openid-4-verifiable-presentations-1_0-20.html#section-6.4-5.2) I believe we should outright reject the request. So I recommend changing the type of `ConstraintsField.filter` to `JSONSchema`, so that the failure happens during the parsing of the `PresentationDefinition`, and there is no error case at this point to handle.

Reply:
I agree with the recommendation to change `filter` type to `JSONSchema`. 

The downside is `JSONSchema` does not implement `serde` traits, which causes some grief for serialization with the `ConstraintsField`. 

For now, I am skipping the serialization/deserialization for the `validator` field, which means that it will show up as `None`, and will require an additional check. 

It would be nice to move the validation error earlier in the construction of the definition, versus runtime.

I've kept the original `validator()` method, but renamed it `try_validator`:

```rust
pub fn try_validator(&self) -> Option<Result<JSONSchema, ValidationError>> {
        self.filter.as_ref().map(JSONSchema::compile)
    }
```

But, I've changed the `validator` method to return the compiled JSONSchema as an Option if it is exists:

```rust
pub fn validator(&self) -> Option<Arc<JSONSchema>> {
        self.validator.clone()
    }
```

In the interest of time and as a temporary solution, I'll change `check_credential_validation` to return a result type and use the `try_validator` method at runtime.

The longer-term solution, which I'd prefer to come in a subsequent PR would be to write a custom serializer/deserializer to handle the `JSONSchema` type, or even the constraints field itself.

Reply:
See: [0ddcdea](https://github.com/spruceid/oid4vp-rs/pull/26/commits/0ddcdea06db7846c1d32e65a94f6888a69770f37)